consul/snapshot/snapshot.go

```diff
@@ -191,3 +191,25 @@ func Restore(logger *log.Logger, in io.Reader, r *raft.Raft) error {
 
 	return nil
 }
+
+// ReadMetadata takes a snapshot from the reader and returns its SnapshotMeta
+func ReadMetadata(in io.Reader) (*raft.SnapshotMeta, error) {
+	// Wrap the reader in a gzip decompressor.
+	decomp, err := gzip.NewReader(in)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decompress snapshot: %v", err)
+	}
+	defer func() {
+		if err := decomp.Close(); err != nil {
+			fmt.Printf("[ERR] snapshot: Failed to close snapshot decompressor: %v", err)
```

Comment:
I kind of struggled with closes like this, too. If we want to log it we'd been to plumb a logger down because this'll just go to stdout without the right formatting. For this, since the decomp is an in-mem thing and we aren't going to leak resources it seems ok to just defer the close without checking (that's what `Verify()` ended up doing).